File: packages/express/src/authdog.ts
import type { PublicKeyPayload } from "@authdog/node-commons";
import type { Request, RequestHandler, Response } from "express";
import { getPublicKeyPayload } from "./commons";
import { logoutHandler } from "./logout";
import { createAttachSession, requireAuth } from "./middleware";
import type { AttachSessionOptions, AuthdogConfig } from "./types";

/** The instance returned by {@link createAuthdog}. */
export interface AuthdogServer {
  /**
   * Middleware that resolves the session and attaches `req.authdog`. Mount it
   * early (typically app-wide) so downstream handlers and `requireAuth` can
   * read the authentication context. Never throws or blocks the request.
   */
  attachSession: (options?: AttachSessionOptions) => RequestHandler;
  /**
   * Gate middleware that returns `401` for unauthenticated requests. This is
   * the real server-side enforcement point — place it before any protected
   * route. Requires `attachSession` to have run first.
   */
  requireAuth: RequestHandler;
  /** Handler that clears the session cookie and performs a safe redirect. */
  logout: (req: Request, res: Response) => void;
  /** The validated, parsed public-key payload (`environmentId`, `identityHost`, …). */
  getPublicKeyPayload: () => PublicKeyPayload;
  /** The validated public-key payload as a JSON string, e.g. to inline into HTML. */
  getPublicKey: () => string;
}

/**
 * Creates an Authdog server instance for Express.
 *
 * The public key is validated and parsed once here — enforcing the trusted
 * identity-host allowlist (SSRF / token-exfiltration protection) — so a
 * malformed or untrusted key fails fast at startup rather than on the first
 * request.
 *
 * ```ts
 * const authdog = createAuthdog({ publicKey: process.env.PK_AUTHDOG! });
 *
 * app.use(authdog.attachSession());
 * app.get("/me", authdog.requireAuth, (req, res) => res.json(req.authdog!.user));
 * app.get("/logout", authdog.logout);
 * ```
 */
export const createAuthdog = (config: AuthdogConfig): AuthdogServer => {
  if (!config.publicKey) {
    throw new Error("Public key is not defined");
  }

  // Validate + parse eagerly so an invalid/untrusted key throws at startup.
  const payload = getPublicKeyPayload(config.publicKey);

  return {
    attachSession: (options?: AttachSessionOptions) =>
      createAttachSession(payload, options),
    requireAuth,
    logout: logoutHandler,
    getPublicKeyPayload: () => payload,
    getPublicKey: () => JSON.stringify(payload),
  };
};

File: packages/go/authdog.go
// Package authdog is the Authdog SDK for Go web backends, with first-class
// middleware for the Gin framework (see gin.go).
//
// It mirrors the TypeScript @authdog/express / @authdog/fastify SDKs on the
// wire — same authdog-session cookie, same OIDC userinfo flow, and the same
// trusted identity-host allowlist — so one Authdog environment serves Node and
// Go services interchangeably.
package authdog

import (
	"errors"
	"net/http"
)

// Config configures a new Authdog instance.
type Config struct {
	// PublicKey is the Authdog public key (pk_…). Required.
	PublicKey string
	// FetchUser controls whether AttachSession performs a userinfo round-trip.
	// Defaults to true. When false, the token is surfaced but IsAuthenticated
	// stays false and the caller is responsible for validating it.
	FetchUser *bool
	// HTTPClient is used for userinfo requests. Defaults to http.DefaultClient.
	HTTPClient *http.Client
}

// Authdog is a configured server instance. The public key is validated and
// parsed once in New — enforcing the trusted identity-host allowlist — so a
// malformed or untrusted key fails fast at startup rather than per request.
type Authdog struct {
	payload   *PublicKeyPayload
	fetchUser bool
	client    *http.Client
}

// New validates the public key and returns an Authdog instance.
func New(cfg Config) (*Authdog, error) {
	if cfg.PublicKey == "" {
		return nil, errors.New("public key is not defined")
	}
	payload, err := ValidateAndParsePublicKey(cfg.PublicKey)
	if err != nil {
		return nil, err
	}
	fetchUser := true
	if cfg.FetchUser != nil {
		fetchUser = *cfg.FetchUser
	}
	client := cfg.HTTPClient
	if client == nil {
		client = http.DefaultClient
	}
	return &Authdog{payload: payload, fetchUser: fetchUser, client: client}, nil
}

// PublicKeyPayload returns the validated, parsed public-key payload.
func (a *Authdog) PublicKeyPayload() PublicKeyPayload { return *a.payload }

// Context is the per-request authentication context resolved by AttachSession.
type Context struct {
	Token           string
	User            any
	IsAuthenticated bool
	UserInfo        *UserInfoResponse
}
